### PR TITLE
 	ir: Track the codegen-reachable items, and use it instead of whitelisted_items() for code generation.

### LINKED ISSUE #834: Behavior change for non-whitelisted types referenced by method
### Input C/C++ Header

```C++
struct T {};
struct U {
  void test(T a);
};
```

### Bindgen Invocation

```
$ bindgen input.hpp --generate types --whitelist-type U
```

### Actual Results

```rust
#[repr(C)]
#[derive(Debug, Copy)]
pub struct U {
    pub _address: u8,
}
#[test]
fn bindgen_test_layout_U() {
    assert_eq!(::std::mem::size_of::<U>() , 1usize , concat ! (
               "Size of: " , stringify ! ( U ) ));
    assert_eq! (::std::mem::align_of::<U>() , 1usize , concat ! (
                "Alignment of " , stringify ! ( U ) ));
}
impl Clone for U {
    fn clone(&self) -> Self { *self }
}
```

### Expected Results

Not really expected result, but before #827, it generates
```rust
#[repr(C)]
#[derive(Debug, Copy)]
pub struct T {
    pub _address: u8,
}
#[test]
fn bindgen_test_layout_T() {
    assert_eq!(::std::mem::size_of::<T>() , 1usize , concat ! (
               "Size of: " , stringify ! ( T ) ));
    assert_eq! (::std::mem::align_of::<T>() , 1usize , concat ! (
                "Alignment of " , stringify ! ( T ) ));
}
impl Clone for T {
    fn clone(&self) -> Self { *self }
}
#[repr(C)]
#[derive(Debug, Copy)]
pub struct U {
    pub _address: u8,
}
#[test]
fn bindgen_test_layout_U() {
    assert_eq!(::std::mem::size_of::<U>() , 1usize , concat ! (
               "Size of: " , stringify ! ( U ) ));
    assert_eq! (::std::mem::align_of::<U>() , 1usize , concat ! (
                "Alignment of " , stringify ! ( U ) ));
}
impl Clone for U {
    fn clone(&self) -> Self { *self }
}
```

### PR DESCRIPTION
**Pull request by emilio**
This standardizes the behavior change at #834, but without regressions.

I've added a few more tests for #833 here.

Closes #834.

### CODE REVIEW

**Comment by emilio** [comment]
(Only the second commit needs review, the other two are #835)

**Comment by fitzgen** [comment]
maybe I should finish reading my inbox before leaving code reviews suggesting exactly what the next item in my inbox does...

File: src/ir/context.rs [hunk 1]
```diff
@@ -1657,16 +1667,43 @@ impl<'ctx> BindgenContext<'ctx> {
...
-            let mut roots: Vec<_> = roots.collect();
             roots.reverse();
+            roots
+        };
 
-            WhitelistedItems::new(self, roots).collect()
-        });
+        // The reversal preserves the expected ordering of traversal, resulting
+        // in more stable-ish bindgen-generated names for anonymous types (like
+        // unions).
```
Comment (fitzgen) [inline src/ir/context.rs]: Why did this comment move here? It should be with the `reverse` call.

File: src/ir/context.rs [hunk 2]
```diff
@@ -1657,16 +1667,43 @@ impl<'ctx> BindgenContext<'ctx> {
...
+                whitelisted_items_predicate,
+            ).collect::<ItemSet>();
+
+        let codegen_items =
+            if self.options().whitelist_recursively {
+                WhitelistedItemsTraversal::new(
+                    self,
+                    roots.clone(),
+                    traversal::codegen_edges,
+                ).collect::<ItemSet>()
```
Comment (fitzgen) [inline src/ir/context.rs]: Mildly surprised the type inference couldn't figure this out.

File: src/codegen/mod.rs [hunk 3]
```diff
@@ -3299,7 +3298,7 @@ pub fn codegen(context: &mut BindgenContext) -> Vec<P<ast::Item>> {
         }
 
         context.resolve_item(context.root_module())
-            .codegen(context, &mut result, whitelisted_items, &());
+            .codegen(context, &mut result, codegen_items, &());
```
Comment (fitzgen) [inline src/codegen/mod.rs]: File a follow up `E-Easy` issue to remove this argument, and remove the `whitelisted_items` parameter from all the codegen functions using `ctx.codegen_items()` instead?

Comment (emilio) [inline src/codegen/mod.rs]: Sure thing.

**Pull request merged**